typing/typecore.ml

```diff
@@ -983,7 +983,24 @@ module Constructor = NameChoice (struct
   let get_name cstr = cstr.cstr_name
   let get_type cstr = cstr.cstr_res
   let lookup_all_from_type loc usage path env =
-    Env.lookup_all_constructors_from_type ~loc usage path env
+    match Env.lookup_all_constructors_from_type ~loc usage path env with
+    | [] ->
+        begin match (Env.find_type path env).type_kind with
+        | Type_open ->
+            (* Extension constructors cannot be found by looking at the type
+               declaration.
+               We scan the whole environment to get an accurate spellchecking
+               hint in the subsequent error message *)
+            let filter lbl =
+              let get_type_path cstr = match (repr (get_type cstr)).desc with
+                | Tconstr(p, _,_ ) -> p
+                | _ -> assert false in
+              compare_type_path env path (get_type_path lbl) in
+            let add_valid x acc = if filter x then (x,ignore)::acc else acc in
+            Env.fold_constructors add_valid None env []
+        | _ -> []
+        end
+    | x -> x
```

Comment:
I think the code would be easier to read if the non-empty case (`( _ :: _) as cstrs`) was handled before the empty case.